### DISCUSSION TITLE
Pipeline forecasters predict logic

### DISCUSSION THREAD

**Discussion by ViktorKaz** [alan-turing-institute/sktime#948] [Q&A] [answered]
Hi @mloning, this refers to PR #856 that I have been working on. You mentioned that the `predict()` method of the pipeline should run through the steps of the pipeline in reverse order. The same logic is implemented in `sktime.forecasting.compose.TransformedTargetForecaster`. We had a short discussion with @fkiraly yesterday and it seems to us that the sequence should be:
1. transform `y` (for all transformers in the pipeline. This is missing from the implementation of `TransformedTargetForecaster`)
1. predict (for the last step of the pipeline, i.e. the forecaster).
1. inverse_transform (for all transformers in reverse order).

Is the first stage of transforming `y` implemented somewhere for `TransformedTargetForecaster` or do you think that the sequence is wrong and at `predict()` the pipeline should only run through the steps in reverse order without first transforming `y`?

**Comment by mloning** [comment, marked as answer]
Hi @ViktorKaz we discussed the sequence of operations here: https://github.com/sktime/enhancement-proposals/blob/main/steps/01_forecasting_api/forecasting_pipeline.md

1. is implemented in `TransformedTargetForecaster`

Note that there is no `y` passed to `predict` when forecasting, so we first need generate the prediction and then inverse transform it.

**Reply by fkiraly** [reply]
yes, I think you are right @mloning - I have to admit I got confused myself there, @ViktorKaz. There isn't anything to transform in `predict`, since `y` is not an input.

**Reply by fkiraly** [reply]
although, in the "network pipeline", transformation of "non-y" arguments would have to happen, e.g., if you are transforming exogeneous variables.

**Reply by ViktorKaz** [reply]
@mloning @fkiraly yes, I understand the point about `predict` not taking `y` as input. I have to admit I missed this point too. 

However, more broadly in the context of Network Pipeline PR #856 the idea that `predict` always iterates through the steps in reverse order is quite limiting. The Network Pipeline was designed principally for directional forecasts. For this use case I think the `predict` sequence will need to be different. However, even for the classic forecasting case which I am working on now, the fact that `predict` iterates though the steps in reverse order can be limiting. For example, I thought that one of the use cases of the Network Pipeline in classical forecasting could be for building ensemble forecasters. For this use case the steps will include `n` forecasters all trained on `y` followed by a final step that applies a weighting or voting rule. For this use case, if at `predict` the steps are reversed, the pipeline will break.

**Reply by mloning** [reply]
For me, the scope of the network pipeline is very unclear. This makes it hard for me to think about the design, implementation and testing of it. So I would start with this question (from #856): What's the simplest use case that is enabled by this PR which is currently not possible in sktime (e.g. with the `TransformedTargetForecaster` or a simple `PipelineForecaster` sketched out in #844)?

**Reply by fkiraly** [reply]
I think there is an established use case for network style pipeliens, and it frequently appears in time seriers: applying transformers to selected variable sets from any number of earlier steps in the pipeline, including output variables and predictions produced by those earlier steps, as well as the original input data. This is beyond capabilities of the "linear" pipeline, which typically always assumes that the variables that are "input" to a step are precisely the output variables of the one earlier step.

**Reply by fkiraly** [reply]
"advanced" use cases include steps that take multiple inputs, like `X`, `Y`, `Z` and the like, to which outputs may be "connected". This becomes especially relevant in annotation use cases, where network pipelines are a differentiator.

**Reply by mloning** [reply]
So what would be a simple example of this which is currently not possible in sktime?
